### ISSUE TITLE
Unify export GUI [$50]

### ISSUE THREAD

**Issue by kintel** [openscad/openscad#649]
Idea to simplify the GUI, reduce mouse clicks and to reduce duplicated code:
- Create one common export dialog with a keyboard shortcut
- Have an option for choosing file format (either through a GUI element or by filename)
- Remember the folder and settings to make the next export operation smoother
- Look at #648 and see if we can add a checkbox for opening the resulting file after export
- Look into whether exporting for 3D printing (octoprint or external service) fits here
- Support for setting export parameters per file format
  - STL: ascii / binary
  - SVG: units, stroke width, fill color, margins (#3270)
  - PDF: color, ruler on/off, meta data (date, openscad version) (#3266), 

- Optional: Configure the toolbar icon to a user-defined export shortcut, instead of just "Export STL"
- Optional: Configure the proposed export hotkey (F7) to a user-defined export shortcut
- Optional: Add echo output (suggested in scope of the PDF export PR #3266).
##

**Comment by t-paul** [comment]
_Idea from #1386 by @kwendenarmo - moved here to collect export enhancements in on place_


When designing complex parts, it is common to divide them into modules. This option is really great! But when you have to export each to print, for example, must be appointed each of these parts as separate files.
It would be great that when only a single module has rendered, the default name for the export file is that of that module.
This would save a lot of time and could eliminate errors as overwrite files by mistake.

**Comment by MichaelAtOz** [comment]
> I use XP mainly and it is as I described.

@nophead you have upgraded haven't you?

**Comment by nophead** [comment]
Yes I use Win7 now. I still don't like the fact that OpenSCAD has started remembering the last folder I saved in. I always want it to start where my source file is. I will probably have to fork OpenSCAD and put it back the way it was.

**Comment by MichaelAtOz** [comment]
>  I will probably have to [fork](https://en.wikipedia.org/wiki/Fork_%28software_development%29) OpenSCAD and put it back the way it was.

I note IIRC you have had a similar perspective on some other issue recently.

It would be a shame to have what I call the fork of death occurring.

As a minimum, you could have a registry (or what ever the Linux/OS X/Qt equivalent is) setting that, at one setting does it your way, and the other the 'other' way.

I realise that, Firefox's about:config for example, can get out of control, but it may be a good approach.

??

**Comment by nophead** [comment]
Could be an option but I can't it makes sense as is. I used to be able to open a file from the recent file list and then do file open to open another file in the same project. Now it defaults to the folder that I last did a file open in, which might have been weeks ago in a different project.

Also I might have two windows open in different projects. I always want the file dialogues to open in the same folder as the file in the editor.

**Comment by GilesBathgate** [comment]
@MichaelAtOz The dreaded "Fear Of Fork". Just make the change in a fork, and if everyone likes it enough, I am sure the maintainer will consider an MR. Especially since the original idea was to "reduce mouse clicks". Certainly opening the last folder reduces the mouse clicks required to get back to that folder ;)

**Comment by nophead** [comment]
Not when the folder the source is in is always where I want it. 

I just looked at a file I download in the temp folder. Switched back to a project I am working on and now it wants to export the STL in the temp folder, bonkers.

**Comment by GilesBathgate** [comment]
@nophead That's what I meant. (corrected my typo). Its quite common for applications to allow you to define a standard "projects" folder in this way. `c:\users\giles\documents\visual studio 2013\Projects` for example. Its certainly less intuitive for the IDE to try and second guess where you would like to go. (and get it wrong)

**Comment by nophead** [comment]
The current behaviour seems to be Save As is always relative to the source file. Open is where the last file open was but not when opened from the recently used list. Export is where the last export was, regardless of file type..

This only seems to make sense if you work in one directory all the time and export to a second directory all the time. It doesn't make sense at all when you have two windows open in different projects or if you often flit between different projects. I move between projects and I also download a lot of code from the forum. Using the last path is almost always wrong and a long way to navigate to where I want it to be. When it used to be relative to the source file it was just one click if I wanted the STLs to be in a subdirectory.

When I build large projects from the command line I use separate directories for STL , PNG and DXF. So a single remembered export path in the GUI is no use at all to me.

The only way to make it sensibly persistent would be to have a map from source file to last export path from that source file, for each export file type. That would require a persistent database. Much simpler to just revert to where the source file is as that is either correct, or close to where the exported file should be, not somewhere completely unrelated like the temp folder or the download folder.

**Comment by GilesBathgate** [comment]
@nophead I agree, I just didn't understand/remember the old behaviour. When you export and STL/AMF by default, it should just take the location of the .scad, and change the extention to .stl/.amf. If you have a sub dir like in your workflow, then indeed its only one click away.

**Comment by nophead** [comment]
Yes that is what it used to do but it got changed some time ago.
